### PR TITLE
Improve documentation for collection associations

### LINKED ISSUE #458: Improve support for collection associations
A few days ago, I run into a problem without pretty solution. I wanted to populate a has many relation with some records, using the same building strategy that was used to build the parent object.

I came out that ugly solution:

``` ruby
factory :offering do
  association :partner, factory: :named_partner
  association :destination, factory: :geolocated_destination

  trait :stay do
    title "Hotel Gran Vía"
    description "Great hotel in a great zone with great views"
    offering_type 'stay'
    price 65
    rooms 70
    stars 4
    event_spaces 3
    photos do
      case @build_strategy
      when FactoryGirl::Strategy::Create then [FactoryGirl.create(:hotel_photo)]
      when FactoryGirl::Strategy::Build then [FactoryGirl.build(:hotel_photo)]
      when FactoryGirl::Strategy::Stub then [FactoryGirl.build_stubbed(:hotel_photo)]
      end
    end
  end
end
```

It works, but is ugly as hell.

After asking in stack overflow (http://stackoverflow.com/questions/13308768/factorygirl-populate-a-has-many-relation-preserving-build-strategy), the best (and only) answer was using `after(:stub)`, `after(:build)` and `after(:create)` callbacks.

``` ruby
factory :offering do
  association :partner, factory: :named_partner
  association :destination, factory: :geolocated_destination

  trait :stay do
    title "Hotel Gran Vía"
    description "Great hotel in a great zone with great views"
    offering_type 'stay'
    price 65
    rooms 70
    stars 4
    event_spaces 3
    after(:stub) do |offering|
      offering.photos = [build_stubbed(:hotel_photo)]
    end
    after(:build) do |offering|
      offering.photos = [build(:hotel_photo)]
    end
    after(:create) do |offering|
      offering.photos = [create(:hotel_photo)]
    end
  end
end
```

Is clearly better, but for me an ideal solution should look like this:

``` ruby
factory :offering do
  association :partner, factory: :named_partner
  association :destination, factory: :geolocated_destination

  trait :stay do
    title "Hotel Gran Vía"
    description "Great hotel in a great zone with great views"
    offering_type 'stay'
    price 65
    rooms 70
    stars 4
    event_spaces 3
    photos { [produce(:hotel_photo)] }
  end
end
```

`#produce` is just an alias to the builder method of the strategy used in the current object.

Anyone thinks that could be a good idea? I'm missing something and there is already a good way to archieve this?

### LINKED ISSUE #1268: Document Evaluator#association
https://github.com/thoughtbot/factory_bot/issues/1063
https://github.com/thoughtbot/factory_bot/issues/1255

The tests in https://github.com/thoughtbot/factory_bot/pull/1419 may serve as a starting point for some code examples as part of this documentation.

### LINKED ISSUE #1304: Edge case for recommended way to define `has_many`
The official way to define `has_many` based on the documentation [here](https://github.com/thoughtbot/factory_bot/blob/master/GETTING_STARTED.md#associations), is by using `after(:create)` callback:

```ruby
FactoryBot.define do
  factory :post do
    body { "nice!" }
  end

  factory :user do
    trait :with_post do
      after(:create) do |user|
        create(:post, user: user)
      end
    end
  end
end
```

Bur if we're doing something to the association before we save the object like:

```ruby
class User < ApplicationRecord
  has_many :posts
  validate :post_validity

  def post_validity
    posts.all? &:valid?
  end
end
```

then in the spec 

```ruby
user = create(:user, :with_post)
user.posts #=> []
```

expectation
```ruby
user.post #=> [Post]
```

I think it was because `user.posts` got cached when `after(:create)` being executed

I can solve it by doing 

```ruby
  factory :user do
    trait :with_post do
      after(:create) do |user|
        create(:post, user: user.reload)
      end
    end
  end
```

but is there any better way?

### PR DESCRIPTION
**Pull request by composerinteralia**
This commit makes `Evaluator#association` and `Evaluator#instance`
officially part of the public API of factory\_bot. Folks were already
using these methods, and all of the examples here were inspired by
examples in the various issues mentioned below.

This commit also adds and updates some tests to match the examples in
the documentation, to ensure that these examples will continue to work
as expected.

Closes #1268 by adding documentation for both `Evaluator#association`
and `Evaluator#instance`. The documentation on interconnected
associations is also relevant to #1063, #1255, and #1309.

Closes #1304 by providing multiple alternative approaches to creating
collection associations, and adding a note about reloading the record in
the callback example. This had also come up back in #549.

Closes #458 by offering `Evaluator#association` as a more flexible way
to build collection associations. This has come up many times over the
years in many different forms, including #426, #487, #640, #1022, #1150,
and #1360.

### CODE REVIEW

File: GETTING_STARTED.md [hunk 1]
```diff
@@ -653,27 +654,51 @@ factory :post do
 
 ### `has_many` associations
 
-Generating data for a `has_many` relationship is a bit more involved,
-depending on the amount of flexibility desired, but here's a surefire example
-of generating associated data.
+There are a few ways to generate data for a `has_many` relationships. The
```
Comment (composerinteralia) [inline GETTING_STARTED.md:657]:
```suggestion
There are a few ways to generate data for a `has_many` relationship. The
```

File: GETTING_STARTED.md [hunk 2]
```diff
@@ -684,71 +709,122 @@ FactoryBot.define do
...
   end
 end
+
+create(:user).posts.length # 0
+create(:user_with_posts).posts.length # 5
+create(:user_with_posts, posts_count: 15).posts.length # 15
 ```
 
-This allows us to do:
+Or for a solution that works with `build`, `build_stubbed`, and `create`
```
Comment (composerinteralia) [inline GETTING_STARTED.md:725]:
```suggestion
Or, for a solution that works with `build`, `build_stubbed`, and `create`
```

File: GETTING_STARTED.md [hunk 3]
```diff
@@ -684,71 +709,122 @@ FactoryBot.define do
...
+  end
+end
+
 create(:user).posts.length # 0
+create(:user_with_posts).posts.length # 1
+build(:user_with_posts).posts.length # 1
+build_stubbed(:user_with_posts).posts.length # 1
+```
+
+Or, for more flexibility you can combine this with the `posts_count` transient
```
Comment (composerinteralia) [inline GETTING_STARTED.md:751]:
```suggestion
For more flexibility you can combine this with the `posts_count` transient
```

File: GETTING_STARTED.md [hunk 4]
```diff
@@ -684,71 +709,122 @@ FactoryBot.define do
...
 
 ```ruby
-FactoryBot.define do
-
-  # language factory with a `belongs_to` association for the profile
-  factory :language do
-    title { "Through the Looking Glass" }
-    profile
+def profile_with_languages(languages_count: 2)
+  FactoryBot.create(:profile) do |user|
```
Comment (composerinteralia) [inline GETTING_STARTED.md:792]:
```suggestion
  FactoryBot.create(:profile) do |profile|
```

File: GETTING_STARTED.md [hunk 5]
```diff
@@ -782,6 +858,58 @@ create(:comment, :for_video)
...
 ```
 
+### Interconnected associations
+
+There are limitless ways records and other objects might be interconnected, and
+factory\_bot may not always be suited to handle those relationships. In some
+cases it makes sense to use factory\_bot to build each individual object, and
+then to write helper methods in plain Ruby to tie those objects together.
+
+That said, come of the more complex relationships can be build in factory\_bot
```
Comment (composerinteralia) [inline GETTING_STARTED.md:868]:
```suggestion
That said, some more complex, interconnected relationships can be built in factory\_bot
```

File: GETTING_STARTED.md [hunk 6]
```diff
@@ -782,6 +858,58 @@ create(:comment, :for_video)
...
+
+class School < ApplicationRecord
+  has_many :students
+  has_many :profiles
+end
+```
+
+We can ensure the student and profile are connected to each other and to the
+same school with a factory like this:
+
```
Comment (composerinteralia) [inline GETTING_STARTED.md:893]:
```suggestion
```

File: GETTING_STARTED.md [hunk 7]
```diff
@@ -782,6 +858,57 @@ create(:comment, :for_video)
 create(:comment, :for_photo)
 ```
 
+### Interconnected associations
+
+There are limitless ways records and other objects might be interconnected, and
```
Comment (composerinteralia) [inline GETTING_STARTED.md:863]:
```suggestion
There are limitless ways objects might be interconnected, and
```
